File: backend/pricing.py
# Pricing Logic for EONITE Configurator

from typing import Tuple

# Base prices per unit by quantity tier
BASE_PRICES = {
    5000: 0.35,
    10000: 0.28,
    25000: 0.22,
    50000: 0.18
}

# Print type cost additions per unit
PRINT_COSTS = {
    "1_color": 0.02,
    "2_colors": 0.04,
    "quadri": 0.08
}

# Size multipliers
SIZE_MULTIPLIERS = {
    "small": 0.8,
    "medium": 1.0,
    "large": 1.3,
    "custom": 1.5
}

# Product category base price adjustments
CATEGORY_ADJUSTMENTS = {
    "sacs_kraft": 1.0,
    "boites": 1.2,
    "luxe": 1.8,
    "gobelets": 0.7,
    "expedition": 0.9
}

def get_quantity_tier(quantity: int) -> int:
    """Get the applicable quantity tier"""
    tiers = sorted(BASE_PRICES.keys())
    applicable_tier = tiers[0]
    for tier in tiers:
        if quantity >= tier:
            applicable_tier = tier
    return applicable_tier

def calculate_unit_price(
    quantity: int,
    print_type: str,
    size: str = "medium",
    category: str = "sacs_kraft"
) -> float:
    """Calculate unit price based on configuration"""
    tier = get_quantity_tier(quantity)
    base_price = BASE_PRICES[tier]
    
    # Add print cost
    print_cost = PRINT_COSTS.get(print_type, 0)
    
    # Apply size multiplier
    size_mult = SIZE_MULTIPLIERS.get(size, 1.0)
    
    # Apply category adjustment
    cat_adj = CATEGORY_ADJUSTMENTS.get(category, 1.0)
    
    unit_price = (base_price + print_cost) * size_mult * cat_adj
    return round(unit_price, 3)

def calculate_total_price(
    quantity: int,
    print_type: str,
    size: str = "medium",
    category: str = "sacs_kraft"
) -> Tuple[float, float]:
    """Calculate total price and unit price"""
    unit_price = calculate_unit_price(quantity, print_type, size, category)
    total_price = unit_price * quantity
    return round(unit_price, 3), round(total_price, 2)

def get_price_breakdown(
    quantity: int,
    print_type: str,
    size: str = "medium",
    category: str = "sacs_kraft"
) -> dict:
    """Get detailed price breakdown for display"""
    tier = get_quantity_tier(quantity)
    base_price = BASE_PRICES[tier]
    print_cost = PRINT_COSTS.get(print_type, 0)
    size_mult = SIZE_MULTIPLIERS.get(size, 1.0)
    cat_adj = CATEGORY_ADJUSTMENTS.get(category, 1.0)
    
    unit_price, total_price = calculate_total_price(quantity, print_type, size, category)
    
    # Calculate savings compared to minimum tier
    max_price = BASE_PRICES[5000]
    potential_unit_price = (max_price + print_cost) * size_mult * cat_adj
    savings_per_unit = potential_unit_price - unit_price
    total_savings = savings_per_unit * quantity if quantity >= 10000 else 0
    
    return {
        "quantity": quantity,
        "quantity_tier": tier,
        "base_price": base_price,
        "print_type": print_type,
        "print_cost": print_cost,
        "size": size,
        "size_multiplier": size_mult,
        "category": category,
        "category_adjustment": cat_adj,
        "unit_price": unit_price,
        "total_price": total_price,
        "savings_per_unit": round(savings_per_unit, 3),
        "total_savings": round(total_savings, 2)
    }

def get_all_tiers_comparison(
    print_type: str,
    size: str = "medium",
    category: str = "sacs_kraft"
) -> list:
    """Get price comparison for all quantity tiers"""
    comparisons = []
    for qty in sorted(BASE_PRICES.keys()):
        unit_price, total_price = calculate_total_price(qty, print_type, size, category)
        comparisons.append({
            "quantity": qty,
            "unit_price": unit_price,
            "total_price": total_price
        })
    return comparisons

File: frontend/src/lib/pricing.js

// Pricing Logic for EONITE Configurator (Ported from Backend)

const BASE_PRICES = {
    5000: 0.35,
    10000: 0.28,
    25000: 0.22,
    50000: 0.18
};

const PRINT_COSTS = {
    "1_color": 0.02,
    "2_colors": 0.04,
    "quadri": 0.08
};

const SIZE_MULTIPLIERS = {
    "small": 0.8,
    "medium": 1.0,
    "large": 1.3,
    "custom": 1.5
};

const CATEGORY_ADJUSTMENTS = {
    "sacs_kraft": 1.0,
    "boites": 1.2,
    "luxe": 1.8,
    "gobelets": 0.7,
    "expedition": 0.9
};

function getQuantityTier(quantity) {
    const tiers = Object.keys(BASE_PRICES).map(Number).sort((a, b) => a - b);
    let applicableTier = tiers[0];
    for (const tier of tiers) {
        if (quantity >= tier) {
            applicableTier = tier;
        }
    }
    return applicableTier;
}

export function calculateTotalPrice(quantity, printType, size = "medium", category = "sacs_kraft") {
    const tier = getQuantityTier(quantity);
    const basePrice = BASE_PRICES[tier];

    const printCost = PRINT_COSTS[printType] || 0;
    const sizeMult = SIZE_MULTIPLIERS[size] || 1.0;
    const catAdj = CATEGORY_ADJUSTMENTS[category] || 1.0;

    const unitPrice = (base_price + printCost) * sizeMult * catAdj;
    const totalPrice = unitPrice * quantity;

    return {
        unitPrice: Number(unitPrice.toFixed(3)),
        totalPrice: Number(totalPrice.toFixed(2))
    };
}

export function getPriceBreakdown(quantity, printType, size = "medium", category = "sacs_kraft") {
    const tier = getQuantityTier(quantity);
    const basePrice = BASE_PRICES[tier];
    const printCost = PRINT_COSTS[printType] || 0;
    const sizeMult = SIZE_MULTIPLIERS[size] || 1.0;
    const catAdj = CATEGORY_ADJUSTMENTS[category] || 1.0;

    const { unitPrice, totalPrice } = calculateTotalPrice(quantity, printType, size, category);

    // Calculate savings
    const maxPrice = BASE_PRICES[5000];
    const potentialUnitPrice = (maxPrice + printCost) * sizeMult * catAdj;
    const savingsPerUnit = potentialUnitPrice - unitPrice;
    const totalSavings = quantity >= 10000 ? savingsPerUnit * quantity : 0;

    return {
        quantity,
        quantity_tier: tier,
        base_price: basePrice,
        print_type: printType,
        print_cost: printCost,
        size,
        size_multiplier: sizeMult,
        category,
        category_adjustment: catAdj,
        unit_price: unitPrice,
        total_price: totalPrice,
        savings_per_unit: Number(savingsPerUnit.toFixed(3)),
        total_savings: Number(totalSavings.toFixed(2))
    };
}

export function getAllTiersComparison(printType, size = "medium", category = "sacs_kraft") {
    const tiers = Object.keys(BASE_PRICES).map(Number).sort((a, b) => a - b);
    return tiers.map(qty => {
        const { unitPrice, totalPrice } = calculateTotalPrice(qty, printType, size, category);
        return {
            quantity: qty,
            unit_price: unitPrice,
            total_price: totalPrice
        };
    });
}
